### ISSUE TITLE
ELECTRA: Pre-training Text Encoders as Discriminators Rather Than Generators

### ISSUE THREAD

**Issue by guchio3** [guchio3/papers#1]
## 概要
Bert の MLM (Masked Language Modeling) より必用計算量・精度両観点で優れた RTD (Replaced Token Detection) を提案し、これを使った finetune により GLUE や SQuAD において従来の 1/4 以下の計算量で RoBERTa や XLNet に匹敵するパフォーマンスを実現。例えば 1 GPU 4 days で GLUE において GPT (30 倍以上の計算量が必要らしい) を上回る。

RTD は各 Token が original の token か generator によって生成された token かを当てるタスクで、bi-directional な機構でも全 token について loss をかけることができるため MLM より効率的 (MLM では [MASK] 部分についてのみしかかけることができなかった)。(ref. figure 2)

![image](https://user-images.githubusercontent.com/18380869/76942782-a4fc2480-6941-11ea-9928-ea5263beb01b.png)

![image](https://user-images.githubusercontent.com/18380869/76942760-9c0b5300-6941-11ea-8514-da284e06fafb.png)

## メモ
 - RTD が GAN like なフレームワークによる学習
     - まず MLM で generator を学習し、このモデルを使って (generator については freeze) discriminator の学習をする
         - GAN のように同時に学習するのは generator 部分に argmax して token を決定する機構が必要なので無理
         - [seq GAN](https://arxiv.org/abs/1609.05473) みたいに強化学習を用いて学習するのもやったらしいけど、MLM で事前学習する形式のほうが良かったらしい
 - generator を smaller にする
     - generator が強すぎる (大きくなると強くなるイメージ) と計算コストが増す & discriminator の解くタスクが難しくなりすぎる
     - layer size を減らし、それ以外の params は同じ
     - discriminator の 1/4~1/2 の大きさが良い

## 感想
 - generator がうまい具合にミスすること前提の学習なので難しそう

## 参照
 - [paper url](https://openreview.net/forum?id=r1xMH1BtvB)
 - [code url](https://github.com/google-research/electra)
 - [SeqGAN: Sequence Generative Adversarial Nets with Policy Gradient](https://arxiv.org/abs/1609.05473)
 - [Adversarial Contrastive Estimation](https://arxiv.org/abs/1805.03642)
      - この論文のフレームワークは GAN というより Adversarial Contrastive Estimation だという話が appendix F にある
